core/policy/repository.go

```diff
@@ -0,0 +1,212 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package policy
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/mysteriumnetwork/node/logconfig/httptrace"
+	"github.com/mysteriumnetwork/node/market"
+	"github.com/mysteriumnetwork/node/requests"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+)
+
+type policyMetadata struct {
+	policy market.AccessPolicy
+	eTag   string
+	rules  market.AccessPolicyRuleSet
+}
+
+// Repository represents async policy fetcher from TrustOracle
+type Repository struct {
+	client *requests.HTTPClient
+
+	policyURL  string
+	policyLock sync.RWMutex
+	policyList []policyMetadata
+
+	fetchInterval time.Duration
+	fetchShutdown chan struct{}
+}
+
+// NewRepository create instance of policy repository
+func NewRepository(client *requests.HTTPClient, policyURL string, interval time.Duration) *Repository {
+	return &Repository{
+		client:        client,
+		policyURL:     policyURL,
+		policyList:    make([]policyMetadata, 0),
+		fetchInterval: interval,
+		fetchShutdown: make(chan struct{}),
+	}
+}
+
+// Start begins fetching policies to repository
+func (pr *Repository) Start() {
+	go pr.fetchLoop()
+}
+
+// Stop ends fetching policies to repository
+func (pr *Repository) Stop() {
+	pr.fetchShutdown <- struct{}{}
+}
+
+// Policy converts given value to valid policy rule
+func (pr *Repository) Policy(policyID string) market.AccessPolicy {
+	policyURL := pr.policyURL
+	if !strings.HasSuffix(policyURL, "/") {
+		policyURL += "/"
+	}
+	return market.AccessPolicy{
+		ID:     policyID,
+		Source: fmt.Sprintf("%v%v", policyURL, policyID),
+	}
+}
+
+// Policies converts given values to list of valid policies
+func (pr *Repository) Policies(policyIDs []string) []market.AccessPolicy {
+	policies := make([]market.AccessPolicy, len(policyIDs))
+	for i, policyID := range policyIDs {
+		policies[i] = pr.Policy(policyID)
+	}
+	return policies
+}
+
+// AddPolicies adds given policy to repository. Also syncs policy rules from TrustOracle
+func (pr *Repository) AddPolicies(policies []market.AccessPolicy) error {
+	pr.policyLock.Lock()
+	defer pr.policyLock.Unlock()
+
+	policyListNew := make([]policyMetadata, len(pr.policyList))
+	copy(policyListNew, pr.policyList)
+
+	for _, policy := range policies {
+		index, exist := pr.getPolicyIndex(policyListNew, policy)
```

Comment:
Maybe it make sense to fetch all active policies at once by providerID? As there is no a lot of policies it would simplify this whole logic a lot. And fetch loop will be fetching all active policies once by one anyway so could fetch in one request too.